include/deal.II/lac/solver_gmres.h

```diff
@@ -1986,16 +2055,68 @@ SolverFGMRES<VectorType>::SolverFGMRES(SolverControl        &cn,
 
 template <typename VectorType>
 DEAL_II_CXX20_REQUIRES(concepts::is_vector_space_vector<VectorType>)
-template <typename MatrixType, typename PreconditionerType>
+template <typename MatrixType, typename... PreconditionerTypes>
 DEAL_II_CXX20_REQUIRES(
   (concepts::is_linear_operator_on<MatrixType, VectorType> &&
-   concepts::is_linear_operator_on<PreconditionerType, VectorType>))
-void SolverFGMRES<VectorType>::solve(const MatrixType         &A,
-                                     VectorType               &x,
-                                     const VectorType         &b,
-                                     const PreconditionerType &preconditioner)
+   (concepts::is_linear_operator_on<PreconditionerTypes, VectorType> && ...)))
+void SolverMPGMRES<VectorType>::solve(
+  const MatrixType &A,
+  VectorType       &x,
+  const VectorType &b,
+  const PreconditionerTypes &...preconditioners)
 {
-  LogStream::Prefix prefix("FGMRES");
+  LogStream::Prefix prefix("MPGMRES");
+  SolverMPGMRES<VectorType>::solve_internal(A, x, b, preconditioners...);
+}
+
+
+
+template <typename VectorType>
+DEAL_II_CXX20_REQUIRES(concepts::is_vector_space_vector<VectorType>)
+template <typename MatrixType, typename... PreconditionerTypes>
+void
+SolverMPGMRES<VectorType>::solve_internal(
+  const MatrixType &A,
+  VectorType       &x,
+  const VectorType &b,
+  const PreconditionerTypes &...preconditioners)
+{
+  constexpr std::size_t n_preconditioners = sizeof...(PreconditionerTypes);
+
+  //
+  // A lambda that cycles through all preconditioners in sequence applying
+  // one to the vector src and storing the result in dst:
+  //
+  const auto preconditioner_vmult = [&](auto &dst, const auto &src) {
+    // We have no preconditioner that we can apply
+    if constexpr (n_preconditioners == 0)
+      {
+        dst = src;
+        return;
+      }
+
+    //
+    // With each invocation of preconditioner_vmult we increase the
+    // variable current_index by one modulo n_preconditioners. We then
+    // cycle through all preconditioners and call the one with a matching
+    // index:
+    //
+
+    static std::size_t current_index = 0;
```

Comment:
Do we need to have a static variable, or can we define `current_index` just at the top of this function and increment it inside the function? The reason is that I would like to use static variables only where really necessary. But overall, this is a very neat implementation.

Reply:
Yes, every `static` variable is not thread-safe unless guarded by a mutex.

Reply:
This is now refactored.

Also please note that this PR is still marked as a draft. So please expect things to be preliminary and to change.